### PR TITLE
Change topojson serialization format

### LINKED ISSUE #1095: Investigate issues loading PA LRC region in staging

### PR DESCRIPTION
**Pull request by maurizi**
## Overview

Upgrading to Node.js 16 in #602 caused a regression in data loading - the maximum size a of string was was reduced from ~1Gb to ~512Mb, which meant we could no longer load our two largest regions by using `JSON.parse(buffer.toString())`

This PR takes a multi-step approach to tackling this problem:
 - I've switched to serialzing the TopoJSON using the v8 binary serialization format: https://github.com/PublicMapping/districtbuilder/commit/c2a0dcb7ff876364e90f534b087b5c91ea9757bc, https://github.com/PublicMapping/districtbuilder/commit/d17738ea29fee0f69c1a90df1e79efec4669dc3e & https://github.com/PublicMapping/districtbuilder/commit/dc52da87c585429013c92c9bd4b7833d107c3f2a
 - Added a command to download the existing TopoJSON files (I used streaming JSON so we can load the now too large files w/o needing to downgrade Node temporarily) https://github.com/PublicMapping/districtbuilder/commit/f8262c278b49a46eb3c1e7e943207ac546b580ec
    - I've already applied this command on all the files used on staging as well as those referenced by `load-dev-data`, but I still need to update production, and any one-off regions will no longer work unless updated
   
After testing this branch on staging, I encountered a lot of resource contention and a large number of layers failed to load the first time, so I added https://github.com/PublicMapping/districtbuilder/commit/807dc9d57910850ca2acb3dd6ba5500b7999d419 & https://github.com/PublicMapping/districtbuilder/commit/e8ea4f103eb3120ea85728ab92e8dbc0e4628c29 to make the `TopologyService` more robust.

### Checklist

- [x] Description of PR is in an appropriate section of `CHANGELOG.md` and grouped with similar changes, if possible

### Demo

Updated healthcheck, showing which layers are loading / pending.
image

### Notes

I was hopeful this would decrease our load times, but it hasn't done so to any meaningful degree. We did buy a lot of headroom in terms of TopoJSON file size - the max file size for [`v8.deserialize` is ~4GB](https://nodejs.org/api/v8.html#v8serializevalue).

## Testing Instructions

Try running `./scripts/manage serialize-topojson ...` locally, and make sure you can run the server

I tested this against staging, and was able to get staging to load successfully.

Closes #1095

### CODE REVIEW

**Review by KlaasH — APPROVED** [review]
Looks good :+1: 
I deleted the `topo.buf` for the NJ config I have in my database, then recreated it and confirmed that it loaded and I could make a map.

Re performance: I found [this comparison of Javascript encoding tools](https://github.com/Adelost/javascript-serialization-benchmark), and [this PR that adds v8.serialize](https://github.com/Adelost/javascript-serialization-benchmark/pull/5) to the mix.  It seems like the tools that are notably faster than `v8.serialize` require a predefined schema, which is probably more trouble than it would be worth for our purposes.

File: src/server/src/districts/services/topology.service.ts [hunk 1]
```diff
@@ -25,25 +61,27 @@ export class TopologyService {
...
-        // Next we load all remaining (unarchived) regions
-        // eslint-disable-next-line
-        for (const s3URI of activeURIs) {
-          this._layers[s3URI] = this.fetchLayer(s3URI, false);
-        }
+        // Load regions a number of regions in parallel, adding as each one completes
+        await asyncLoop(
+          sortedRegions.map(region => () => {
+            const promise = this.fetchLayer(region.s3URI, region.archived);
+            this._layers = { ...this._layers, [region.s3URI]: promise };
```
[the remark this answers is not in the document]
Comment (maurizi) [reply src/server/src/districts/services/topology.service.ts:80]: Yeah I tried / failed at this as well - good to know it wasn't just me 😅

File: src/server/src/districts/services/topology.service.ts [hunk 2]
```diff
@@ -25,25 +61,27 @@ export class TopologyService {
...
+          !STATE_ORDER.includes(region.regionCode),
+          STATE_ORDER.indexOf(region.regionCode)
+        ]);
 
-        // Next we load all remaining (unarchived) regions
-        // eslint-disable-next-line
-        for (const s3URI of activeURIs) {
-          this._layers[s3URI] = this.fetchLayer(s3URI, false);
-        }
+        // Load regions a number of regions in parallel, adding as each one completes
```
Comment (maurizi) [inline src/server/src/districts/services/topology.service.ts:76]:
```suggestion
        // Load a number of regions in parallel, adding another as each one completes
```

File: src/manage/src/commands/serialize-topojson.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+export default class SerializeTopojson extends Command {
+  static description = `reprocess topojson files into binary format
+  
+  Pass a list of s3_uri paths to reprocess, e.g.
+  serialize-topojson s3://bucket-name/regions/US/PA s3://other-bucket-name/regions/US/DE
+`;
+
+  static strict = false;
+
+  // Streaming reader for GeoJSON files. Works on files over 1GB, but is slow.
```
Comment (maurizi) [inline src/manage/src/commands/serialize-topojson.ts:20]:
```suggestion
  // Streaming reader for GeoJSON files. Works on files over 512MB, but is slow.
```

File: src/server/src/districts/services/topology.service.ts [hunk 4]
```diff
@@ -11,8 +14,41 @@ import { GeoUnitProperties } from "../entities/geo-unit-properties.entity";
 import _ from "lodash";
 import { getObject, s3Options } from "../../common/s3-wrapper";
 
+const MAX_RETRIES = 5;
+// Loading a topojson layer is a mix of I/O and CPU intensive work,
+// so we can afford to have more concurrent than we have cores, but not too many
```
Comment (maurizi) [inline src/server/src/districts/services/topology.service.ts:19]:
```suggestion
// so we can afford to have more layers loading than we have cores, but not too many
```

**Pull request merged**